Get current user

hannahkamundson opened the pull request:
Only look at filterSecurity, serializers, and view. The rest is Ivan's changes.

This is taking in a request to get the current user. For now I'm leaving the first user too so we can actually see what we are doing

HTempleman commented on api/gsn_api/gsndb/views.py:
```diff
@@ -1,155 +1,459 @@
 from django.contrib.contenttypes.models import ContentType
+from django.utils import timezone
+from django.http import HttpResponseRedirect
+from django.db.models import Q
```
Very cool feature that I was unaware of!

HTempleman commented on api/gsn_api/gsndb/views.py:
```diff
@@ -1,155 +1,459 @@
+
+        Note: We override the response given by the post_note() method here
+        and instead explicitly redirect.
+        """
```
It doesn't appear that you are overriding the post_note response. The note should be amended to reflect this.

HTempleman commented on api/gsn_api/gsndb/views.py:
```diff
@@ -1,155 +1,459 @@
+            return Response({"Sorry": "this user does not have access to edit this referral for this student."})
+        else:
+            note_data = {
+                "user": user.get_user().id,
```
shouldn't 'note_data' be called 'referral_data'?

hannahkamundson commented:
@HTempleman Nice catch with the ProgramList. Added that. Also added the note on top of serializers to explain what is going on. Also changed from note to referral. Pushed the changes, let me know what you think.

Regardless of whether or not the code is in a method, the FilterUser would still have to be called. My initial feelings are that I don't see a benefit to putting it into a method as it isn't ever reused? Maybe you have a different idea for how it would be reused? Let me know!

HTempleman commented:
I chose my words poorly when I suggested that you create a method. Rather, I believe that the code block I mentioned above should be contained within a standalone function, similar to how post_note is a standalone function. By doing this, you create a single point of failure, which is easier for debugging. For example:

```python 
def check_security(request):
    user = FilterSecurity(request)
    if access_level == user.get_my_access():
        queryset = user.get_my_schools()
    elif access_level == user.get_all_access():
        queryset = user.get_accessible_schools()
   return queryset
```

By capturing the logic in a single function, you can avoid easy mistakes, such as the absent `user = FilterSecurity(request)` line from ProgramList. Using the above logic, you could simply include the following line in ProgramList and other classes whose querysets are filtered by permissions:

```python
queryset = check_security(request)
```
I'm sure that you could alternatively weave that logic into a method for FilterSecurity, which would likely be the most elegant approach. Whichever approach you choose, it is important that we keep code [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) to keep bugs from sprouting up, as was the case in ProgramList.

hannahkamundson commented:
Hm, I'm still struggling to see a huge benefit from that considering the repeated code is just the if/else criteria. Yes we would be reducing code, but to what degree? We would still have to pass in consequent. Additionally, the bug still would have occurred if we make it a method in FilterSecurity because FilterSecurity would still have to be initialized.

That being said, absolutely feel free to rewrite it as a method in FilterSecurity or even a function like post_note! I don't feel committed to my code. Are we able to push it through for now and you can pull it from master and refactor whenever you have time?

HTempleman commented:
We can merge the changes as is. I'll pull things in when I'm back at my house.
